### PR TITLE
Pagerando guarded optimizers

### PR DESCRIPTION
**Pull request by yln**
Only add Pagerando passes to compiler pipeline if we actually compiling with Pagerando activated

### CODE REVIEW

**Review by rinon — APPROVED** [review]

File: lib/CodeGen/TargetPassConfig.cpp [hunk 1]
```diff
@@ -607,7 +607,8 @@ void TargetPassConfig::addMachinePasses() {
     addPass(&LocalStackSlotAllocationID, false);
   }
 
-  addPass(createPagerandoBinningPass());
+  if (TM->isPagerando())
+    addPass(createPagerandoBinningPass());
```
Comment (yln) [inline lib/CodeGen/TargetPassConfig.cpp:611]: Although PagerandoBinning is target-agnostic, we could push this inside the target-specific `PassConfig`s to have the everything Pagerando right next to each other. What do you think?

File: lib/Target/AArch64/AArch64TargetMachine.cpp [hunk 2]
```diff
@@ -495,7 +495,7 @@ bool AArch64PassConfig::addILPOpts() {
 }
 
 void AArch64PassConfig::addPreRegAlloc() {
-  if (TM->getOptLevel() != CodeGenOpt::None) { // TODO(yln): Maybe factor out the TM->getOptLevel() != CodeGenOpt::None check
+  if (TM->getOptLevel() != CodeGenOpt::None && TM->isPagerando()) {
```
Comment (yln) [inline lib/Target/AArch64/AArch64TargetMachine.cpp:498]: Push this to the end of `addPreRegAlloc` so it's even closer to register allocation?

File: lib/CodeGen/TargetPassConfig.cpp (hunk 1 above)
Comment (rinon) [reply lib/CodeGen/TargetPassConfig.cpp:611]: I think it's not ideal to duplicate that for each target. Let's see what LLVM devs prefer.

File: lib/Target/AArch64/AArch64TargetMachine.cpp (hunk 2 above)
Comment (rinon) [reply lib/Target/AArch64/AArch64TargetMachine.cpp:498]: I don't think the positioning of this matters, unless we are also moving binning. If we do move binning to target-specific pass management, then we should move it.

**Pull request merged**